File: thisweek_masuipeo/q608/tyama_codeiq608_iter.php
#!/usr/bin/php
<?php
function reverse(&$a,$start,$size){
	$end=$start+$size-1;
	for(;$start<$end;$start++){
		$z=$a[$start];$a[$start]=$a[$end];$a[$end]=$z;
		$end--;
	}
}
function unique_permutation($a,$_n=null){
	$siz=count($a);
	$n=$_n ?? $siz;
	if($n<0||$siz<$n)return;
	$i=0;
	for(;;){
		yield array_slice($a, 0, $n);
		reverse($a,$n,$siz-$n);
		for($i=$siz-2;$i>=0;$i--)if($a[$i]<$a[$i+1])break;
		if($i<0){
			//reverse($a,0,$siz);
			return;
		}
		$k=$i;
		for($i=$siz-1;$i>=$k+1;$i--)if($a[$k]<$a[$i])break;
		$l=$i;
		$z=$a[$k];$a[$k]=$a[$l];$a[$l]=$z;
		reverse($a,$k+1,$siz-($k+1));
	}
}

$N=6;
$e0=array_fill(0,$N*2,0);$f0=array_fill(0,$N*2,0);$i=0;$r=0;
for($i=0;$i<$N;$i++)$e0[$N+$i]=$f0[$N+$i]=1;
$e=array(0);
$f=array(0);
foreach(unique_permutation($e0) as $e_){
	for($i=0;$i<$N*2;$i++)$e[$i+1]=$e[$i]+$e_[$i];
	foreach(unique_permutation($f0) as $f_){
		for($i=0;$i<$N*2;$i++){
			$f[$i+1]=$f[$i]+$f_[$i];
			if($e[$i]==$f[$i]&&$e[$i+1]==$f[$i+1])break;
		}
		if($i==$N*2)$r++;
	}
}
echo $r.PHP_EOL;
?>

File: thisweek_masuipeo/q608/tyama_codeiq608_iter.jl
#!/usr/bin/env julia

function reverse(a,start::Int,size::Int)
	local end_=start+size-1
	while start<end_
		local z=a[start]
		a[start]=a[end_]
		a[end_]=z
		end_-=1
		start+=1
	end
end

function next_permutation(a0,n::Int)
	local ch = Channel{typeof(a0)}(8)
	function _producer()
		if n<0||length(a0)<n
			close(ch)
			return
		end
		local a=a0[:]
		while true
			put!(ch,a[1:n])
			reverse(a,n+1,length(a)-n)
			local i=length(a)-1
			while i>0
				if a[i]<a[i+1]
					break
				end
				i-=1
			end
			if i<=0
				#reverse(a,1,length(a))
				break
			end
			local k=i
			i=length(a)
			while i>=k
				if a[k]<a[i]
					break
				end
				i-=1
			end
			local l=i
			local z=a[k]
			a[k]=a[l]
			a[l]=z
			reverse(a,k+1,length(a)+1-(k+1))
		end
		close(ch)
	end
	@async _producer()
	return ch
end

Base.@ccallable function julia_main(args::Vector{String})::Cint
	local N=6
	local e0=Array{Int}(undef,N*2)
	local f0=Array{Int}(undef,N*2)
	local i=1
	local r=0
	while i<=N
		e0[N+i]=1
		f0[N+i]=1
		i+=1
	end
	sort!(e0)
	sort!(f0)
	local e=Array{Int}(undef,N*2+1)
	local f=Array{Int}(undef,N*2+1)
	for e_ in next_permutation(e0,length(e0))
		for i in 1:N*2
			e[i+1]=e[i]+e_[i]
		end
		for f_ in next_permutation(f0,length(f0))
			i=1
			while i<=N*2
				f[i+1]=f[i]+f_[i]
				if e[i]==f[i]&&e[i+1]==f[i+1]
					break
				end
				i+=1
			end
			if i>N*2
				r+=1
			end
		end
	end
	println(r)
	return 0
end

if get(ENV, "COMPILE_STATIC", "false") == "false"
    julia_main(ARGS)
end
